### PR TITLE
[Concurrency] Allow synchronous functions to be a subtype of 'async' functions

### PR DESCRIPTION
**Pull request by DougGregor**
Implement support for synchronous functions being a subtype of 'async' functions to line up with changes to the async/await proposal. This includes:

* An implicit conversion from a synchronous function to its `async` counterpart
* The ability for a synchronous function to witness an `async` protocol requirement

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please clean test Windows

File: include/swift/AST/DiagnosticsSema.def [hunk 1]
```diff
@@ -551,6 +551,10 @@ ERROR(throws_functiontype_mismatch,none,
       "invalid conversion from throwing function of type %0 to "
       "non-throwing function type %1", (Type, Type))
 
+ERROR(async_functiontype_mismatch,none,
+      "invalid conversion from 'async' function of type %0 to "
+      "synchronous function type %1", (Type, Type))
```
Comment (slavapestov) [inline include/swift/AST/DiagnosticsSema.def:556]: "of type" maybe?

**Review by slavapestov — APPROVED** [review]

File: include/swift/AST/DiagnosticsSema.def (hunk 1 above)
Comment (DougGregor) [reply include/swift/AST/DiagnosticsSema.def:556]: I sorts like it this way? We trying to convert to a synchronous function type

**Pull request merged**